storage: log reasons for splits and merges

tbg opened the pull request:
This will make it easier to reason about clashes between the merge and
split queues, which we're supposedly seeing.

My confidence in load-based splitting is rather low because there
doesn't seem to be any testing outside of basic unit tests. I added a
check to the split queue that seems reasonable but was omitted: before
going through with a load-based split, verify that the load is still
high enough to support it. Previously, it seems that the split would've
been carried out even if there hadn't been a write to the replica for a
long time. (This would only have manifested itself in production with
very bursty workloads plus a busy split queue).

This new logging will allow us to identify questionable splits as well
as thrashing between the split and merge queues (along with the reasons
for which these splits and merges occur).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34008)
<!-- Reviewable:end -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34008#-:-LWGm-8V977P4itJVdEc:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/storage/replica_command.go, line 173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34008#-LWGlJNAA_bCv8vokwtE:-LWGlJNBCd6XtEvUiwwY:bdn3dil) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6b03a7e05f62d78a29a3855795c632fde0f707e/pkg/storage/replica_command.go#L173)):*
> ```Go
> 		}
> 
> 		reply, lastErr = r.adminSplitWithDescriptor(ctx, args, r.Desc(), true /* delayable */, "manual")
> ```

Shouldn't you pass `reason` instead of `"manual"` here?


<!-- Sent from Reviewable.io -->

tbg committed:
storage: log reasons for splits and merges

This will make it easier to reason about clashes between the merge and
split queues, which we're supposedly seeing.

My confidence in load-based splitting is rather low because there
doesn't seem to be any testing outside of basic unit tests. I added a
check to the split queue that seems reasonable but was omitted: before
going through with a load-based split, verify that the load is still
high enough to support it. Previously, it seems that the split would've
been carried out even if there hadn't been a write to the replica for a
long time. (This would only have manifested itself in production with
very bursty workloads plus a busy split queue).

This new logging will allow us to identify questionable splits as well
as thrashing between the split and merge queues (along with the reasons
for which these splits and merges occur).

Release note: None

tbg commented:
Thanks for catching that @petermattis. Maybe that's why I didn't see any load based splits (though there definitely isn't a test that intentionally triggers them, or at least not one that's reasonably easy to find).

bors r=petermattis
